catalyst/utils/distributed.py

```diff
@@ -0,0 +1,237 @@
+from collections import OrderedDict
+import copy
+import inspect
+import os
+import socket
+import subprocess
+import sys
+from typing import Dict, Tuple
+
+import torch
+from torch import nn
+import torch.distributed
+
+from catalyst import utils
+from catalyst.utils.tools.typing import (
+    Criterion, Device, Model, Optimizer, Scheduler
+)
+
+
+def get_rank() -> int:
+    """
+    If torch.distributed is initialized returns ``rank``, otherwise ``-1``
+    :return:
+    """
+    if torch.distributed.is_initialized():
+        return torch.distributed.get_rank()
+    else:
+        return -1
+
+
+def is_apex_available() -> bool:
+    try:
+        import apex  # noqa: F401
+        from apex import amp  # noqa: F401
+        return True
+    except ImportError:
+        return False
+
+
+def assert_fp16_available() -> None:
+    """
+    Asserts for installed and available Apex FP16
+    """
+    assert torch.backends.cudnn.enabled, \
+        "fp16 mode requires cudnn backend to be enabled."
+
+    assert is_apex_available(), "NVidia Apex package must be installed. "  \
+                                "See https://github.com/NVIDIA/apex."
+
+
+def distributed_mean(value: float):
+    if torch.distributed.is_initialized():
+        value = torch.tensor(
+            value,
+            dtype=torch.float,
+            device=f"cuda:{torch.cuda.current_device()}",
+            requires_grad=False
+        )
+        torch.distributed.all_reduce(value)
+        value = float(value.item() / torch.distributed.get_world_size())
+    return value
+
+
+def get_default_params(fn, exclude=None):
+    """
+
+    :param fn:
+    :param exclude:
+    :return:
+    """
+    argspec = inspect.getfullargspec(fn)
+    default_params = zip(
+        argspec.args[-len(argspec.defaults):], argspec.defaults
+    )
+    if exclude is not None:
+        default_params = filter(lambda x: x[0] not in exclude, default_params)
+    default_params = dict(default_params)
+    return default_params
+
+
+def get_slurm_params():
+    cmd = "scontrol show hostnames '%s'" % os.environ["SLURM_JOB_NODELIST"]
+    nodes = subprocess.getoutput(cmd).split()
+    num_nodes = int(os.environ["SLURM_JOB_NUM_NODES"])
+    current_node = os.environ["SLURMD_NODENAME"]
+    master_node = socket.gethostbyname(nodes[0])
+    cur_node_idx = nodes.index(current_node)
+    return cur_node_idx, num_nodes, master_node
+
+
+def is_slurm():
```

Comment:
following `is_apex_available` what do you think about `is_slurm_available`?

Reply:
done